Update order modifier queries

JPry opened the pull request:
### 🎫 Ticket

[ET-2268]
<!-- Ticket ID, if there's any put it between brackets -->

### 🗒️ Description

This PR makes the following changes to the order modifier database queries:

1. It makes correct use of the `QueryBuilder` objects to create the queries instead of relying on direct strings
2. Some of the main queires now allow for more parameters, including:
  - `limit` - limit the number of results returned, with a default of 10.
  - `order` - specify ordering ASC or DESC.
  - `orderby` - specify what column to order the results by. Defaults to `id`.
  - `offset` - specify how many results to skip before returning results.
  - `page` - specify what page of results to retrieve. Overrides `offset`.
  - `status` - specify an array of statuses for the results
3. The Fees admin table uses the pagination features of the updated queries
4. There are tests in place to ensure the correct results are provided.

### 🎥 Artifacts <!-- if applicable-->
<!-- 🎥 screencast(s) or 📷 screenshot(s) -->

### ✔️ Checklist
- [ ] Ran `npm run changelog` to add changelog file(s). More info [here](https://docs.theeventscalendar.com/developer/git/changelogs/#process)
- [x] Code is covered by **NEW** `wpunit` or `integration` tests.
- [x] Code is covered by **EXISTING** `wpunit` or `integration` tests.
- [ ] Are all the **required** tests passing?
- [ ] Automated code review comments are addressed.
- [ ] Have you added Artifacts?
- [x] Check the base branch for your PR.
- [ ] Add your PR to the project board for the release.

dpanta94 commented on tests/order_modifiers_integration/Repositories/Fees.php:
```diff
@@ -0,0 +1,80 @@
+
+		// Test that we get the correct count of all fees.
+		$count = $repo->get_search_count();
+		$this->assertEquals( 20, $count );
```
I see 5 modifiers being inserted twice, so 10. why is the assertion for 20 ?

JPry replied:
I intended to add more, good catch.

dpanta94 commented on tests/order_modifiers_integration/Repositories/Fees.php:
```diff
@@ -0,0 +1,80 @@
+	 * @test
+	 * @return void
+	 */
+	public function should_find_all_fees() {
```
Are we sure that the most crucial functionality introduced is covered by this test case ? If not, could we become please ?

JPry replied:
I added some more tests, and I plan to add even more in a follow-up.

Camwyn reviewed (changes requested):
Dimi brought up some other things in Slack, so I'm not going to cover that here.

JPry commented on src/Tickets/Commerce/Order_Modifiers/Table_Views/Order_Modifier_Table.php:
```diff
@@ -104,39 +104,34 @@ public function __construct(
-				'order'         => $order,
-				'modifier_type' => $this->modifier->get_modifier_type(),
+				'search_term' => $search,
+				'orderby'     => sanitize_text_field( tribe_get_request_var( 'orderby', 'display_name' ) ),
+				'order'       => sanitize_text_field( tribe_get_request_var( 'order', 'asc' ) ),
```
Thanks, I'll change those. This was just moving these statements around, but I'm glad it revealed that we should be doing something differently here.

Camwyn commented on src/Tickets/Commerce/Order_Modifiers/Table_Views/Order_Modifier_Table.php:
```diff
@@ -104,39 +104,34 @@ public function __construct(
-		// Set the items for the table.
-		$this->items = $data;
+		// Get the total number of items.
+		$total_items = $this->modifier->find_count_by_search( [ 'search_term' => $search ] );
```
Interesting.

Since we're doing a limit query and then an "unlimited" one - can we just do one unlimited one and return a limited set?

JPry replied:
I'm up for doing it that way instead.

I don't know if it would make any difference, but I considered trying to return only a single column instead of all of the results, but I don't know if that would make a substantial difference.

Camwyn replied:
What you have can do for now but we should explore this a bit more when we're not over deadline.

dpanta94 commented on tests/order_modifiers_integration/Coupons/Create_Coupon_Modifiers_Test.php:
```diff
@@ -26,6 +26,9 @@ class Create_Coupon_Modifiers_Test extends Order_Modifiers_TestCase {
 	 */
 	public function does_table_render_correctly() {
+		// Skip this test until the coupon feature is enabled.
+		$this->markTestSkipped( 'Pending the coupon feature being enabled.' );
```
Why skip ? This was working. If this now fails, there could be a valid reason.

Camwyn commented on src/Tickets/Commerce/Order_Modifiers/Table_Views/Order_Modifier_Table.php:
```diff
@@ -146,6 +141,63 @@ public function prepare_items() {
+
+		// Get the total number of items.
+		if ( count( $this->items ) < $per_page && $current_page === 1 ) {
+			$total_items = count( $this->items );
```
can just return here

Camwyn replied:
...unless we're going to filter the value?

Camwyn commented on src/Tickets/Commerce/Order_Modifiers/Table_Views/Order_Modifier_Table.php:
```diff
@@ -146,6 +141,63 @@ public function prepare_items() {
+		if ( count( $this->items ) < $per_page && $current_page === 1 ) {
+			$total_items = count( $this->items );
+		} else {
+			$total_items = $this->modifier->find_count_by_search(
```
then we can return here - without the else...

Camwyn commented on src/Tickets/Commerce/Order_Modifiers/Table_Views/Order_Modifier_Table.php:
```diff
@@ -146,6 +141,63 @@ public function prepare_items() {
+
+		// Get the total number of items.
+		if ( count( $this->items ) < $per_page && $current_page === 1 ) {
+			$total_items = count( $this->items );
```
same as above on the returning

Camwyn commented on src/Tickets/Commerce/Order_Modifiers/Repositories/Order_Modifiers.php:
```diff
@@ -185,46 +188,45 @@ public function find_by_id( int $id ): Order_Modifier {
 		// Merge passed arguments with defaults.
-		$args = array_merge( $defaults, $args );
+		$args       = wp_parse_args( $args, $this->get_default_query_params() );
+		$valid_args = $this->get_valid_params( $args );
```
Not huge, but better to use one term - `params` or `args` (in this case...`args`?)

JPry commented:
@Camwyn @dpanta94 This is ready for another look

From slack:

> query 48 is used to count ALL the fees => that should not happen if the primary query returns less than 10 fees (or whatever the limit is)

That was a good idea, and I updated the code to behave this way. The second query is skipped when the number of fees retrieved is less than the limit.

> queries 49-52 are then individual meta queries to determine if each fee is applied per ticket or to all tickets => that should be avoided and it will be more performant if we grab that info with our primary query using a JOIN

Another good idea, I updated the code to perform the JOIN.

> as a result, current code would perform 12 queries in every page reload in order to display 10 fees.
> 12 = primary + count + 1 meta query for each fee
> when it could do
> 1 if fees less than limit or 2 if fees retrieved are equal to the limit

The code has been updated so that it will perform only one query if less than the limit, and 2 if it meets the limit.

The only exception is searching, but that is something we can add on in the next iteration.
